EIPS/eip-7701.md

```diff
@@ -40,382 +43,121 @@ Transactions of this type are referred to as "AA transactions".
-3. `TXPARAMCOPY` - dynamic formula
-
-    ```
-    minimum_word_size = (size + 31) / 32
-    static_gas = 3
-    dynamic_gas = 3 * minimum_word_size + memory_expansion_cost
-    ```
-
-#### Time range validity (optional)
-
-If the `valid_until` field is non-zero, the transaction is only valid for inclusion in a block with a timestamp at most `valid_until` value.
-Similarly, the transaction is only valid for inclusion in blocks with a timestamp at most the `valid_after` value.
-
-#### Sender Deployment (optional)
-
-Inputs to the `deployer` contract are not defined by the protocol and are controlled by the `deployer_data` parameter.
-
-The sender deployment frame MUST result in the `sender` address becoming initialized with contract code.
-
-This step is performed with the `role_sender_deployment` role.
-
-#### Sender Validation
+### `CURRENT_ROLE` and `ACCEPT_ROLE` opcodes
 
-This step is performed with the `role_sender_validation` role.
+`current_frame_role` is a context variable set by the AA transaction to the current role.
+During AA transactions it is set to the current role in the transaction's lifecycle for each top level call. 
+During non-AA transactions it is always set to `ROLE_SENDER_EXECUTION`.
+It remains unchanged on `DELEGATECALL` but is reset to `ROLE_SENDER_EXECUTION` on `CALL` / `STATICCALL` / `CALLCODE`. This behavior resembles `msg.sender`.
 
-In order for the transaction to be considered valid, the
-sender validation frame MUST return without reverting.
+The `CURRENT_ROLE` opcode returns the `current_frame_role` value.
 
-#### Paymaster Validation (optional)
+The `ACCEPT_ROLE` opcode is equivalent to `RETURN` in the sense that it copies a memory slice, ends execution, and pastes the memory slice onto parent returndata and parent memory, and except:
 
-This step is performed with the `role_paymaster_validation` role.
+* It accepts the `frame_role` as the additional input parameter, and reverts if it differs from the `current_frame_role`
 
-In order for the transaction to be considered valid, the paymaster validation frame MUST return without reverting.
-If it does, the `Paymaster` contract is charged for the transaction gas costs instead of the `Sender`.
+For each `role` in the transaction's lifecycle, a successful `ACCEPT_ROLE` is expected with the `frame_role == role`.
+If any validation frame failed to perform an `ACCEPT_ROLE` matching its role, the transaction fails the validity checks and cannot be included.
 
-#### Sender Execution
+## `TXPARAM*` opcodes
 
-This step is performed with the `role_sender_execution` role.
+The `TXPARAMDLOAD`, `TXPARAMSIZE`, `TXPARAMCOPY` opcodes follow the pattern of `CALLDATA*` / `RETURNDATA*` opcode families.
 
-Inputs to the `Sender` contract are not defined by the protocol and are controlled by the `sender_execution_data` parameter.
 
-#### Paymaster post-operation frame (optional)
+Each `TXPARAM*` opcode takes an extra stack input value compared to its `CALLDATA*` equivalent.
```

Comment:
```suggestion
Each `TXPARAM*` opcode takes an index as an extra stack input value compared to its `CALLDATA*` equivalent:

| Stack      | Value        |
| ---------- | ------------ |
| `top - 0`  | `destOffset`  |
| `top - 1`  | `offset`     |
| `top - 2`  | `size`     |
| `top - 3`  | `index`     |
```

Reply:
I understand that the idea is to remove the complexity of tables - I pushed a rephrase.